File: code/collections.cs
var datas = new List&lt;SensorData> 
{
    new SensorData { Id = 1, Location = "A", Value = 2.89 },
    new SensorData { Id = 2, Location = "B", Value = 12.01 },
    new SensorData { Id = 3, Location = "B", Value = 11.89 },
    new SensorData { Id = 4, Location = "A", Value = 3.11 },
    new SensorData { Id = 5, Location = "A", Value = -456.0 }
};

var avgs = datas
            .Where(e => e.Value > -50.0)
            .GroupBy(e => e.Location)
            .Select(g => new { 
                    Location = g.Key,
                    Value = g.Average(e => e.Value) });

// { Location: A, Value: 3.0 }
// { Location: B, Value: 11.95 }

File: code/collections.kt
var datas = listOf(
    SensorData(1, "A", 2.89),
    SensorData(2, "B", 12.01),
    SensorData(3, "B", 11.89),
    SensorData(4, "A", 3.11),
    SensorData(5, "A", -456.0)
)

var avgs = datas
    .filter { it.value > -50.0 }
    .groupBy(SensorData::location)
    .map { Location(it.key, it.value.map(SensorData::value).average()) }
    
// (location=A, value=3.0)
// (location=B, value=11.95)